tests/codegen/float/f16.rs

```diff
@@ -0,0 +1,131 @@
+// Verify that our intrinsics generate the correct LLVM calls for f16
+
+#![crate_type = "lib"]
+#![feature(f16)]
+#![feature(core_intrinsics)]
+
+// CHECK-LABEL: i1 @f16_eq(
+#[no_mangle]
+pub fn f16_eq(a: f16, b: f16) -> bool {
+    // CHECK: fcmp oeq half %{{.+}}, %{{.+}}
+    a == b
+}
+
+// CHECK-LABEL: i1 @f16_ne(
+#[no_mangle]
+pub fn f16_ne(a: f16, b: f16) -> bool {
+    // CHECK: fcmp une half %{{.+}}, %{{.+}}
+    a != b
+}
+
+// CHECK-LABEL: i1 @f16_gt(
+#[no_mangle]
+pub fn f16_gt(a: f16, b: f16) -> bool {
+    // CHECK: fcmp ogt half %{{.+}}, %{{.+}}
+    a > b
+}
+
+// CHECK-LABEL: i1 @f16_ge(
+#[no_mangle]
+pub fn f16_ge(a: f16, b: f16) -> bool {
+    // CHECK: fcmp oge half %{{.+}}, %{{.+}}
+    a >= b
+}
+
+// CHECK-LABEL: i1 @f16_lt(
+#[no_mangle]
+pub fn f16_lt(a: f16, b: f16) -> bool {
+    // CHECK: fcmp olt half %{{.+}}, %{{.+}}
+    a < b
+}
+
+// CHECK-LABEL: i1 @f16_le(
+#[no_mangle]
+pub fn f16_le(a: f16, b: f16) -> bool {
+    // CHECK: fcmp ole half %{{.+}}, %{{.+}}
+    a <= b
+}
+
+// CHECK-LABEL: half @f16_neg(
+#[no_mangle]
+pub fn f16_neg(a: f16) -> f16 {
+    // CHECK: fneg half %{{.+}}
+    -a
+}
+
+/* todo: clang extends then truncates f16 math. We just emit fadd-style ops. Is this ok? */
+
+// CHECK-LABEL: half @f16_add(
+#[no_mangle]
+pub fn f16_add(a: f16, b: f16) -> f16 {
+    // CHECK: fadd half %{{.+}}, %{{.+}}
+    a + b
+}
```

Comment:
I don't think there is any problem with the `todo` here but just want to check. Code compiles successfully and math has been correct in my run-pass tests (not included in this PR)

Reply:
It winds up with the same assembly as clang,  calling `__extendhfsf2`, on each argument then `__truncsfhf2` on the result https://llvm.godbolt.org/z/cGx3K4vhW. Does this mean that we will need to use `compiler_builtins` rather than the LLVM intrinsic?

Reply:
I actually went ahead and checked this: for the specific case of `f16` and `f32` extending and truncating happens to work out because `f32` has sufficient precious to avoid any double-rounding issues, at least for basic operations such as add, sub, mul, div, sqrt. So this is probably fine as it is.

Reply:
Thanks for looking. With the modules there are a tests that should hopefully catch any sort of precision errors, but that will be in my next PR.